src/clr-angular/data/datagrid/datagrid-items.ts

```diff
@@ -3,50 +3,62 @@
  * This software is released under MIT license.
  * The full license information can be found in LICENSE in the root directory of this project.
  */
+import {NgForOf, NgForOfContext} from "@angular/common";
 import {
     Directive,
     DoCheck,
     Input,
     IterableDiffer,
     IterableDiffers,
     OnChanges,
+    SimpleChange,
     SimpleChanges,
     TemplateRef,
-    TrackByFunction
+    TrackByFunction,
+    ViewContainerRef
 } from "@angular/core";
 
 import {Items} from "./providers/items";
 
 @Directive({
     selector: "[clrDgItems][clrDgItemsOf]",
 })
-export class ClrDatagridItems implements OnChanges, DoCheck {
-    private _rawItems: any[];
+export class ClrDatagridItems<T> implements OnChanges, DoCheck {
+    private iterableProxy: NgForOf<T>;
+    private _rawItems: T[];
+    private _differ: IterableDiffer<T>|null = null;
+
     @Input("clrDgItemsOf")
-    public set rawItems(items: any[]) {
-        this._rawItems = items ? items : [];
+    public set rawItems(items: T[]) {
+        this._rawItems = items ? items : [];  // local copy for ngOnChange diffing
+        this._items.all = this._rawItems;     // initial copy for the iterableProxy
     }
-    private _differ: IterableDiffer<any>;
 
-    constructor(public template: TemplateRef<any>, private _differs: IterableDiffers, private _items: Items) {
-        _items.smartenUp();
+    @Input("clrDgItemsTrackBy")
+    set trackBy(value: TrackByFunction<T>) {
+        this.iterableProxy.ngForTrackBy = value;
     }
 
-    ngOnChanges(changes: SimpleChanges): void {
-        if ("rawItems" in changes) {
-            const currentItems = changes.rawItems.currentValue;
-            if (!this._differ && currentItems) {
-                this._differ = this._differs.find(currentItems).create(this._items.trackBy);
-            }
-        }
+    constructor(public template: TemplateRef<NgForOfContext<T>>, private differs: IterableDiffers,
+                private _items: Items, private vcr: ViewContainerRef, private _differs: IterableDiffers) {
+        _items.smartenUp();
+        this.iterableProxy = new NgForOf<T>(this.vcr, this.template, this.differs);
+        _items.change.subscribe(items => {
+            this.iterableProxy.ngForOf = items;
+            this.iterableProxy.ngDoCheck();
+        });
     }
 
-    @Input("clrDgItemsTrackBy")
-    set trackBy(value: TrackByFunction<any>) {
-        this._items.trackBy = value;
+    ngOnChanges(changes: SimpleChanges): void {
+        // TODO: REMOVE WHEN THIS IS UPGRADED TO ANGULAR 6.1 (consumers will have compile errors)
```

Comment:
When we do this, does that mean we'll require Angular 6.1+? Thats kind of a breaking change?

Reply:
Since it is acting as a proxy for `NgForOf` this note is there to remind me to remove it when the Clarity Angular version moves ahead of angular 6.1. I believe it that is the version this was added: https://github.com/angular/angular/pull/23378 and I see it in the `6.1.0-beta.0` tag: https://github.com/angular/angular/commits/6.1.0-beta.0